### PR TITLE
📝(project) add base documentation

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Now that the project is used in production, we need to explain key concepts and document commands and backends.

## Proposal

- [x] integrate mkdocs
- [x] automate documentation build in the CI
- [x] document backends
- [x] document commands (should be automated given command docstrings)
- [x] ~write tutorial~ postponed (will be handled in another PR)
- [x] improve developer user guide

### CODE REVIEW

File: .env.dist [hunk 1]
```diff
@@ -22,16 +22,16 @@ RALPH_APP_DIR=/app/.ralph
...
-# RALPH_SWIFT_OS_IDENTITY_API_VERSION=
-# RALPH_SWIFT_OS_USER_DOMAIN_NAME=
-# RALPH_SWIFT_OS_PROJECT_DOMAIN_NAME=
-# RALPH_SWIFT_OS_TENANT_ID=
-# RALPH_SWIFT_OS_TENANT_NAME=
-# RALPH_SWIFT_OS_USERNAME=
-# RALPH_SWIFT_OS_PASSWORD=
-# RALPH_SWIFT_OS_REGION_NAME=
-# RALPH_SWIFT_OS_STORAGE_URL=
+# RALPH_SWIFT_OS_AUTH_URL=http://swift:35357/v3/
```
Comment (SergioSim) [inline .env.dist:25]: Good Idea to put the test values in the .env.dist, thanks for that)

File: Makefile [hunk 2]
```diff
@@ -93,11 +106,23 @@ logs: ## display app logs (follow mode)
 	@$(COMPOSE) logs -f app
 .PHONY: logs
 
-run: ## starts supported backends servers
```
Comment (SergioSim) [inline Makefile:96]: There are still some references for this command in the Makefile:

At line 65 we have `es-index: run`
At line 136 we have `test: run`

File: README.md [hunk 3]
```diff
@@ -1,77 +1,41 @@
...
-### Tray development
-
-Ralph is distributed along with its tray (a deployable package for Kubernetes
-clusters using [Arnold](https://github.com/openfun/arnold)). If you intend to
-work on this tray, please refer to Arnold's documentation.
-
-### Local Swift Storage development
-
-To make ralph work with a local instance of Swift:
+> You should sustitute `COMMAND` by the target command, _e.g._ `list`, to see
```
Comment (SergioSim) [inline README.md:30]:
```suggestion
> You should substitute `COMMAND` by the target command, _e.g._ `list`, to see
```

File: docs/backends.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+Elasticsearch backend is mostly used for indexation purpose (as a datalake) but
+it can also be used to fetch indexed data from it.
+
+#### Backend parameters
+
+Elasticsearch backend parameters required to connect to a cluster are:
+
+- `hosts`: a list of cluster hosts to connect to (_e.g._ `["http://elasticsearch-node:9200"]`)
+- `index`: the elasticsearch index where to get/put documents
+- `client_options`: all options from the offical client are supported and can
```
Comment (SergioSim) [inline docs/backends.md:128]:
```suggestion
- `client_options`: all options from the official client are supported and can
```

File: docs/contribute.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+To start playing with `ralph`, you should build it using the `bootstrap` Make target:
+
+```
+$ make bootstrap
+```
+
+Once the project has been bootstrapped, you may want to edit generated `.env`
+file to set up available backends parameters that will be injected in the
+running container as environmnet variables to configure Ralph (see [backends
```
Comment (SergioSim) [inline docs/contribute.md:13]:
```suggestion
running container as environment variables to configure Ralph (see [backends
```

File: docs/index.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Alternatively, Docker users can pull the latest ralph image and start playing
+with it:
+
+```sh
+# Pull latest docker image and get usage
+$ docker run --rm -i fundocker/ralph:latest ralph --help
+
+# Pro tip: define an alias to ease your life
+$ alias ralph="docker run --rm -i fundocker/ralph:latest ralph"
```
Comment (SergioSim) [inline docs/index.md:57]: I like this tip :+1:

File: docs/index.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+[available commands](./commands.md).
+
+## Contributing
+
+This project is intended to be community-driven, so please, do not hesitate to
+get in touch if you have any question related to our implementation or design
+decisions.
+
+We try to raise our code quality standards and expect contributors to follow
+the recommandations from our
```
Comment (SergioSim) [inline docs/index.md:70]:
```suggestion
the recommendations from our
```

File: docs/models/index.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
+Ralph can be considered versatile since at some point it is
+log-content-agnostic: most commands will work as expected without transforming
+logged events content or format.
+
+But on the other side, Ralph is a tool dedicated to learning events processing
+so we've implemented key features related to learning events validation and
+convertion. For now, we mostly focus on two learning events standards: [Open
```
Comment (SergioSim) [inline docs/models/index.md:7]:
```suggestion
conversion. For now, we mostly focus on two learning events standards: [Open
```

**Review by SergioSim — COMMENTED** [review]
Beautiful Documentation! Thanks a lot Julien)

File: Makefile (hunk 2 above)
Comment (jmaupetit) [reply Makefile:96]: Good catch! :pray:

**Comment by jmaupetit** [comment]
Thank you for your feedback @SergioSim :heart:

File: README.md [hunk 9]
```diff
@@ -1,77 +1,41 @@
-# Ralph, a learning analytics processor to feed your LRS
+# Ralph
 
-## Getting started with development
+Ralph, is a command-line tool to fetch, extract, convert and push your tracking
```
Comment (quitterie-lcs) [inline README.md:3]:
```suggestion
Ralph is a command-line tool to fetch, extract, convert and push your tracking
```

File: docs/backends.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+format using this backend.
+
+> For now the LDP backend is **read-only** as we consider that it is mostly
+> used to collect primary logs and not as a Ralph target. Feel free to get in
+> touch to prove us wrong, or better: submit your proposal for the `write`
+> method implementation.
+
+#### Backend parameters
+
+To access to OVH's LDP API, you need to register Ralph as an authorized
```
Comment (quitterie-lcs) [inline docs/backends.md:54]:
```suggestion
To access OVH's LDP API, you need to register Ralph as an authorized
```

File: docs/contribute.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+# Start Elasticsearch local backend
+$ make run-es
+# Start Swift local backend
+$ make run-swift
+# Start all local backends
+$ make run-all
+```
+
+Now that you have started at least the `elasticsearch` and `swift` backends,
+it's time to play with those:
```
Comment (quitterie-lcs) [inline docs/contribute.md:88]:
```suggestion
it's time to play with them:
```

File: docs/index.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
+# Introduction
+
+Ralph, is a command-line tool to fetch, extract, convert and push your tracking
```
Comment (quitterie-lcs) [inline docs/index.md:3]:
```suggestion
Ralph is a command-line tool to fetch, extract, convert and push your tracking
```

File: docs/models/index.md [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+log-content-agnostic: most commands will work as expected without transforming
+logged events content or format.
+
+But on the other side, Ralph is a tool dedicated to learning events processing
+so we've implemented key features related to learning events validation and
+conversion. For now, we mostly focus on two learning events standards: [Open
+edX](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/analytics.html)
+and [xAPI](https://adlnet.gov/projects/xapi/).
+
+Data validation and serialisation/de-serialisation is achieved using
```
Comment (quitterie-lcs) [inline docs/models/index.md:11]:
```suggestion
Data validation and serialisation/de-serialisation are achieved using
```

**Review by quitterie-lcs — COMMENTED** [review]
Congrats for this comprehensive work! 
There is still a very very small issue that comes from the automated update for `command` and `model` documentations: 

- sometimes, there is the description sentence with a point (for example in Open edx events > base > baseContextModel)
 
> Represents the base context model inherited by all event contexts. 

- and sometimes not (for example in ralph > extract)

> Extract input events from a container format using a dedicated parser

File: docs/models/index.md [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+[pydantics](https://pydantic-docs.helpmanual.io) models that are documented in
+the following subsections:
```
Comment (quitterie-lcs) [inline docs/models/index.md:13]:
```suggestion
the following subsection:
```

Comment (quitterie-lcs) [reply docs/models/index.md:13]: As there is, for now, only one bullet point

Comment (jmaupetit) [reply docs/models/index.md:13]: Yes. I anticipated a little bit :wink:

**Comment by jmaupetit** [comment]
>  There is still a very very small issue that comes from the automated update for command and model documentations

You are totally right, we need to stay consistent. IMO this should be addressed in a new PR.

**Comment by jmaupetit** [comment]
@SergioSim & @quitterie-lcs I've addressed all your comments in separated fixup commits. Is there any chance that you have  time for a last review (and validation)? :pray:

File: docs/index.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+Ralph is distributed as a python package and a Docker image.
+
+> If you choose to install `ralph` in your native environment (without using
+> Docker), please make sure that **Python 3.9** is installed (and your default
+> python distribution).
+
+Ralph package can be installed from PyPI using the `pip` tool:
+
+```sh
+# Create a new virtualenv (optionnal)
```
Comment (quitterie-lcs) [inline docs/index.md:38]: @jmaupetit I just forgot this mistake when reviewing... 
```suggestion
# Create a new virtualenv (optional)
```

**Review by quitterie-lcs — APPROVED** [review]
Let's go!

**Pull request merged**